gobblin-core/src/main/java/gobblin/writer/SimpleDataWriterBuilder.java

```diff
@@ -34,7 +49,33 @@
    */
   @Override
   public DataWriter<byte[]> build() throws IOException {
-    return new SimpleDataWriter(this, this.destination.getProperties());
+    buildEncoders();
+    return new SimpleDataWriter(this, encoders, this.destination.getProperties());
+  }
+
+  private void buildEncoders() {
+    // Compress first since compressing encrypted data will give no benefit
+    Map<String, Object> compressionConfig =
+        CompressionConfigParser.getConfigForBranch(this.destination.getProperties(), this.branches, this.branch);
+    if (compressionConfig != null) {
+      encoders.add(CompressionFactory.buildStreamCompressor(compressionConfig));
+    }
+
+    Map<String, Object> encryptionConfig =
+        EncryptionConfigParser.getConfigForBranch(this.destination.getProperties(), this.branches, this.branch);
+    if (encryptionConfig != null) {
+      encoders.add(EncryptionFactory.buildStreamEncryptor(encryptionConfig));
+    }
   }
 
+  @Override
+  public String getFileName(State properties) {
+    StringBuilder filenameBuilder = new StringBuilder(super.getFileName(properties));
+    for (StreamCodec codec : encoders) {
+      filenameBuilder.append('.');
+      filenameBuilder.append(codec.getTag());
```

Comment:
You could add the additional extensions in the `Writer` and avoid the weirdness, can't you?